corehq/apps/sso/backends.py

```diff
@@ -0,0 +1,56 @@
+from django.contrib.auth.backends import ModelBackend
+from django.contrib.auth.models import User
+
+from corehq.apps.sso.models import IdentityProvider, AuthenticatedEmailDomain
+
+
+class SsoBackend(ModelBackend):
+    """
+    Authenticates against an IdentityProvider and SAML2 session data.
+    """
+
+    def authenticate(self, request, username=None, idp_slug=None,
+                     is_handshake_successful=False, **kwargs):
+        if not (username and idp_slug and is_handshake_successful):
```

Comment:
Should probably also check `request` here since it could theoretically be `None`:
> request is an `HttpRequest` and may be `None` if it wasn’t provided to `authenticate()` (which passes it on to the backend).

https://docs.djangoproject.com/en/2.2/topics/auth/customizing/#writing-an-authentication-backend

In practice this is currently not super important since we always supply a `request` when calling `auth.authenticate()` in a way that will match this backend. However, it is more future proof to do the check.